modulus/distributed/utils.py

```diff
@@ -400,36 +403,48 @@ def gather_v_wrapper(
     if comm_size == 1:
         return tensor
 
-    gather_list = [None] * comm_size
     tensor_shape = list(tensor.shape)
+    x_recv = [None] * comm_size
+    x_send = [None] * comm_size
 
     for r in range(comm_size):
-        tensor_shape[dim] = sizes[r]
-        gather_list[r] = torch.empty(
+        if rank == dst:
+            tensor_shape[dim] = sizes[r]
+        else:
+            tensor_shape[dim] = 0
+
+        x_recv[r] = torch.empty(
             tensor_shape,
             dtype=tensor.dtype,
             device=tensor.device,
         )
 
-    # dist.scatter doesn't support tensors of different shape
-    # so this implementation is using explicit send/recv combinations
-    if rank == dst:
-        req_list = [None] * comm_size
-        for r in range(comm_size):
-            if r == dst:
-                gather_list[r] = tensor
-            else:
-                req_list[r] = dist.irecv(gather_list[r], src=r, group=group)
+        if r == dst:
+            x_send[r] = tensor
+        else:
+            tensor_shape[dim] = 0
+            x_send[r] = torch.empty(
+                tensor_shape,
+                dtype=tensor.dtype,
+                device=tensor.device,
+            )
 
-        for r in range(comm_size):
-            if r != dst:
-                req_list[r].wait()
+    dist.all_to_all(x_recv, x_send, group=group)
```

Comment:
Can this be simplified by just using [`dist.all_to_all_single`](https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_to_all_single)?

Reply:
see comment above w.r.t. all_to_all_single being marked as experimental and subject to change at the moment